### ISSUE TITLE
rake tasks fails to run ruby v2.5.0 / rails v5.2.0.beta2

### ISSUE THREAD

**Issue by neumayr** [Shopify/bootsnap#114]
`rake` fails to run with latest ruby v2.5.0 / rails v5.2.0.beta2

> `already initialized constant Rake::*`

##### System 

```
$ rvm --version
rvm 1.29.3 (latest)
$ ruby -v
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-darwin17]
$ gem --version
2.7.4
$ rake --version
rake, version 12.3.0
$ gem list | grep bootsnap
bootsnap (1.1.7)
```

<details>
<summary>Output of <code>rake -T</code></summary>

```
$ rake -T
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:3: warning: already initialized constant Rake::VERSION
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:3: warning: previous definition of VERSION was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::MAJOR
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of MAJOR was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::MINOR
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of MINOR was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::BUILD
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of BUILD was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::OTHER
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of OTHER was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:8: warning: already initialized constant Rake::Version::NUMBERS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:8: warning: previous definition of NUMBERS was here
WARNING: Possible conflict with Rake extension: String#ext already exists
WARNING: Possible conflict with Rake extension: String#pathmap already exists
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/linked_list.rb:110: warning: already initialized constant Rake::LinkedList::EMPTY
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/linked_list.rb:110: warning: previous definition of EMPTY was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/scope.rb:41: warning: already initialized constant Rake::Scope::EMPTY
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/scope.rb:41: warning: previous definition of EMPTY was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils.rb:10: warning: already initialized constant FileUtils::RUBY
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils.rb:10: warning: previous definition of RUBY was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils.rb:109: warning: already initialized constant FileUtils::LN_SUPPORTED
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils.rb:109: warning: previous definition of LN_SUPPORTED was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:17: warning: already initialized constant Rake::FileUtilsExt::DEFAULT
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:17: warning: previous definition of DEFAULT was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:44: warning: already initialized constant Rake::FileList::ARRAY_METHODS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:44: warning: previous definition of ARRAY_METHODS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:47: warning: already initialized constant Rake::FileList::MUST_DEFINE
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:47: warning: previous definition of MUST_DEFINE was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:51: warning: already initialized constant Rake::FileList::MUST_NOT_DEFINE
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:51: warning: previous definition of MUST_NOT_DEFINE was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:55: warning: already initialized constant Rake::FileList::SPECIAL_RETURN
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:55: warning: previous definition of SPECIAL_RETURN was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:61: warning: already initialized constant Rake::FileList::DELEGATING_METHODS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:61: warning: previous definition of DELEGATING_METHODS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:86: warning: already initialized constant Rake::FileList::GLOB_PATTERN
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:86: warning: previous definition of GLOB_PATTERN was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:381: warning: already initialized constant Rake::FileList::DEFAULT_IGNORE_PATTERNS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:381: warning: previous definition of DEFAULT_IGNORE_PATTERNS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:387: warning: already initialized constant Rake::FileList::DEFAULT_IGNORE_PROCS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:387: warning: previous definition of DEFAULT_IGNORE_PROCS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/promise.rb:12: warning: already initialized constant Rake::Promise::NOT_SET
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/promise.rb:12: warning: previous definition of NOT_SET was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/application.rb:13: warning: already initialized constant Rake::CommandLineOptionError
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:13: warning: previous definition of CommandLineOptionError was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/application.rb:41: warning: already initialized constant Rake::Application::DEFAULT_RAKEFILES
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:41: warning: previous definition of DEFAULT_RAKEFILES was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/task_arguments.rb:108: warning: already initialized constant Rake::EMPTY_TASK_ARGS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task_arguments.rb:108: warning: previous definition of EMPTY_TASK_ARGS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/invocation_chain.rb:55: warning: already initialized constant Rake::InvocationChain::EMPTY
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/invocation_chain.rb:55: warning: previous definition of EMPTY was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/early_time.rb:21: warning: already initialized constant Rake::EARLY
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/early_time.rb:21: warning: previous definition of EARLY was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/late_time.rb:17: warning: already initialized constant Rake::LATE
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/late_time.rb:17: warning: previous definition of LATE was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:4: warning: already initialized constant Rake::Backtrace::SYS_KEYS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:4: warning: previous definition of SYS_KEYS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:5: warning: already initialized constant Rake::Backtrace::SYS_PATHS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:5: warning: previous definition of SYS_PATHS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:8: warning: already initialized constant Rake::Backtrace::SUPPRESSED_PATHS
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:8: warning: previous definition of SUPPRESSED_PATHS was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:12: warning: already initialized constant Rake::Backtrace::SUPPRESSED_PATHS_RE
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:12: warning: previous definition of SUPPRESSED_PATHS_RE was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:16: warning: already initialized constant Rake::Backtrace::SUPPRESS_PATTERN
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:16: warning: previous definition of SUPPRESS_PATTERN was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake.rb:70: warning: already initialized constant FileList
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake.rb:70: warning: previous definition of FileList was here
/Users/dev/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake.rb:71: warning: already initialized constant RakeFileUtils
/Users/dev/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake.rb:71: warning: previous definition of RakeFileUtils was here
```
</details>

##### Workarounds
* `rails -T` is working fine.
* If I comment `require 'bootsnap/setup'` in `config/boot.rb`, 
  `rake -T` runs without errors

**Comment by BroiSatse** [comment]
Having the same issue with ruby 2.5.0 / rails 5.1.4. Downgrading to ruby 2.4.3 removes all the warnings and fixes the issue. Also, running `bin/rake` instead of `rake` seems to work as well - `bin/rake` requires `config/boot` (which requires bootsnap) before loading `rake`, so it seems to be some dependency requirement order issue.

My initial guess was that rake is loading `rake` before the bootsnap is loaded and that somehow causes bootsnap to re-require same file again. Need to dig more into that though.

**Comment by nynhex** [comment]
I had the same issue and followed @BroiSatse suggestion which solved the issue temporarily for me.

**Comment by ojab** [comment]
Can you please check if https://github.com/Shopify/bootsnap/pull/136 fixes the issue? You can try it with `gem 'bootsnap', github: 'ojab/bootsnap', require: false` in Gemfile.

**Comment by neumayr** [comment]
> Can you please check if #136 fixes the issue? You can try it with `gem 'bootsnap', github: 'ojab/bootsnap', require: false` in Gemfile.

@ojab Yes, with your bootsnap fork `rake -T` is working again

**Comment by burke** [comment]
I bet this is fixed in version 1.2.0.pre, please let me know whether that fixes this issue or not.

**Comment by neumayr** [comment]
> I bet this is fixed in version 1.2.0.pre, please let me know whether that fixes this issue or not.

@burke :no_good: Nope, `v1.2.0.pre` isn't fixing the problem.

**Comment by burke** [comment]
1.3.0 includes @ojab's patch, so I'm closing this. Thanks everyone!
